refiner-lib/src/tx_hash_tracker.rs

```diff
@@ -0,0 +1,257 @@
+use aurora_refiner_types::{near_block::NEARBlock, near_primitives::hash::CryptoHash};
+use std::path::Path;
+
+/// A helper object for tracking the NEAR transaction hash that caused each NEAR receipt
+/// to be produced (potentially indirectly via a number of other receipts). The main interface
+/// includes the `get_tx_hash` function to query the helper for the transaction hash associated
+/// with a given receipt hash, and two functions to mutate the tracker's state:
+/// `consume_near_block` and `on_block_end`. The purpose of `consume_near_block` is to update
+/// the tracker's state with the new receipts that are created in that block. The purpose of
+/// `on_block_end` is to give the tracker an opportunity to prune away old state that is no
+/// longer needed since the block at the given height is finished processing.
+pub struct TxHashTracker {
+    inner: TxHashTrackerImpl,
+}
+
+impl TxHashTracker {
+    pub fn new<P: AsRef<Path>>(storage_path: P, start_height: u64) -> anyhow::Result<Self> {
+        let inner = TxHashTrackerImpl::new(storage_path, start_height)?;
+        Ok(Self { inner })
+    }
+
+    pub fn get_tx_hash(&mut self, rx_hash: &CryptoHash) -> Option<CryptoHash> {
+        self.inner.get_tx_hash(rx_hash)
+    }
+
+    pub fn consume_near_block(&mut self, near_block: &NEARBlock) -> anyhow::Result<()> {
+        let block_height = near_block.block.header.height;
+
+        let tx_iter = near_block
+            .shards
+            .iter()
+            .filter_map(|s| s.chunk.as_ref())
+            .flat_map(|c| c.transactions.iter());
+
+        // Track receipts created from transactions
+        for tx in tx_iter {
+            let tx_hash = tx.transaction.hash;
+            for rx_hash in tx.outcome.execution_outcome.outcome.receipt_ids.iter() {
+                self.inner.record_rx(*rx_hash, tx_hash, block_height)?;
+            }
+        }
+
+        let rx_iter = near_block
+            .shards
+            .iter()
+            .flat_map(|s| s.receipt_execution_outcomes.iter());
+
+        // Track receipts created from other receipts
+        for rx in rx_iter {
+            let rx_hash = &rx.receipt.receipt_id;
+            let tx_hash = match self.get_tx_hash(rx_hash) {
+                Some(tx_hash) => tx_hash,
+                None => {
+                    tracing::warn!("Transaction provenance unknown for receipt {}", rx_hash);
+                    continue;
+                }
+            };
+            for rx_hash in rx.execution_outcome.outcome.receipt_ids.iter() {
+                self.inner.record_rx(*rx_hash, tx_hash, block_height)?;
```

Comment:
Reading about rocksdb I founded a batch write method. I can't find info about performance but batch operations should be better. This is the api link in case you want to take a look.
https://docs.rs/rocksdb/0.10.0/rocksdb/struct.WriteBatch.html

Reply:
Yeah, good call! I was using the batched interface for pruning, but missed doing it for the inserts. Done in https://github.com/aurora-is-near/borealis-engine-lib/pull/31/commits/33d941baf740e0d752cd0aeb146c51cff6de7946